README
DHCP-Test version 1.000
=======================

See README.md for usage instructions

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  Exporter::Tidy
  IO::Interface

COPYRIGHT AND LICENCE

Copyright (C) 2021 by Ton Hospel

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself, either Perl version 5.32.1 or,
at your option, any later version of Perl 5 you may have available.


Changes
Revision history for Perl extension DHCP::Test.

1.000	2021-06-05 19:29:31 UT

0.01  Sat Mar 27 11:40:10 2021
	- original version; created by h2xs 1.23 with options
		-AXn DHCP::Test


lib/DHCP/Test/Package.pm
package DHCP::Test::Package;
# $HeadURL: xxx/lib/DHCP/Test/Package.pm $
# $Id: xxx $

# START HISTORY
# autogenerated by release_pm
use strict;
use warnings;
use vars qw($VERSION $SUB_VERSION $release_time %history);
$VERSION = "1.000";
$SUB_VERSION = "027";
$release_time = 1622921371;	## no critic (UselessNoCritic MagicNumbers)
%history = (
  'Changes' => {
    '1.000' => '1.000'
  },
  'MANIFEST' => {
    '1.000' => '1.000'
  },
  'MANIFEST.SKIP' => {
    '1.000' => '1.000'
  },
  'Makefile.PL' => {
    '1.000' => '1.000'
  },
  'README' => {
    '1.000' => '1.000'
  },
  'README.md' => {
    '1.000' => '1.000'
  },
  'bin/dhcp_test' => {
    '1.000' => '1.000'
  },
  'lib/DHCP/Test.pm' => {
    '1.000' => '1.000'
  },
  'lib/DHCP/Test/Package.pm' => {
    '1.000' => '1.000'
  },
  't/00_syntax.t' => {
    '1.000' => '1.000'
  },
  'version_check' => {
    '1.000' => '1.000'
  }
);

use Carp;

my $epoch_base;

sub SUB_VERSION {
    return $SUB_VERSION;
}

sub FULL_VERSION {
    return "$VERSION.$SUB_VERSION";
}

sub release_time {
    if (!defined $epoch_base) {
	require Time::Local;
	$epoch_base = Time::Local::timegm(0,0,0,1,0,1970);	## no critic (UselessNoCritic MagicNumbers)
    }
    return $release_time + $epoch_base;
}

sub released {
    my ($package, $version) = @_;
    my $p = $package;
    $p =~ s{::}{/}g;
    my $history = $history{"lib/$p.pm"} ||
	croak "Could not find a history for package '$package'";
    my $lowest = 9**9**9;
    for my $v (keys %$history) {
	$lowest = $v if $v >= $version && $v < $lowest;
    }
    croak "No known version '$version' of package '$package'" if
	$lowest == 9**9**9;
    return $history->{$lowest};
}
1;
__END__

=for stopwords DHCP::Test globals

=head1 NAME

DHCP::Test::Package - Version and history of DHCP::Test

=head1 SYNOPSIS

  use DHCP::Test::Package;

  $epoch_time = DHCP::Test::Package->release_time();

  $package_version = DHCP::Test::Package->VERSION;

  $min_package_version = DHCP::Test::Package::released($module, $module_version);

=head1 DESCRIPTION

In the context of this documentation a C<package> is a set of files that
together make up a perl extension, not to be confused with the more normal
perl concept of a namespace for globals.

A package release with a certain package version number contains a number of
modules and other files with their own version numbers. This module contains a
history of which files with which versions where in which package release and
also knows when the current package was released.

This module contains a few simple methods to query this.

The version number of this Package.pm module is always equal to the version
number of the package release. This means you can use this to query this
release number and also makes it a convenient target for Makefile.PL
dependencies.

=head1 METHODS

=over

=item X<release_time>$epoch_time = DHCP::Test::Package->release_time()

Returns the the number of non-leap seconds since whatever time the system that
calls this function considers to be the epoch the last time
L<release_pm|release_pm(1)> was run (even if that was on another system with a
different epoch). Since the idea is to run L<release_pm|release_pm(1)> just
before releasing the package this is therefore the value of
L<time()|perlfunc/time> on the system calling this function at the moment the
package was released. This number is suitable for feeding to
L<gmtime()|perlfunc/gmtime> and L<local_time()|perlfunc/local_time>

=item X<VERSION>$package_version = DHCP::Test::Package->VERSION

This is the normal L<UNIVERSAL::VERSION|UNIVERSAL/VERSION> method you can use
on all modules, but this particular module is guaranteed to have a version
number that is the same as that of the package

=item X<released>$min_package_version = DHCP::Test::Package::released($module, $module_version)

Given a module name (e.g. Foo::Bar) and a module version number (e.g. 1.023)
returns the lowest package version in which that module (as the file
F<Foo/Bar.pm>) had at least that module version

=back

=head1 EXPORTS

None.

=head1 SEE ALSO

L<release_pm|release_pm(1)> which can be used to keep Package.pm files up to
date.

=cut

# END HISTORY


Makefile.PL
#!/usr/bin/perl -w
use 5.008001;
use strict;
use warnings;
use ExtUtils::MakeMaker;
use FindBin;

our $VERSION = '1.000';

# Allows to suppress all program installation with -n (library only)
use Getopt::Std;
our ($opt_n, $opt_y);
getopts("ny") || die "Usage: $0 [-n] [-y]\n";

my @programs_to_install;

if (!$opt_n || $opt_y) {
    print <<EOT	## no critic (UselessNoCritic InputOutput::RequireCheckedSyscalls)
Note that you can avoid the following questions by passing
the '-n' or '-y' option to 'Makefile.PL'.

EOT
        if !$opt_y;

    # Only do this site_perl substitution for non binary extensions!
    my $site_lib = "$ExtUtils::MakeMaker::Config{siteprefix}/lib/site_perl";
    if ($ExtUtils::MakeMaker::Config{installsitelib} ne $site_lib &&
        grep $_ eq $site_lib, @INC) {
        print <<EOT	## no critic (UselessNoCritic InputOutput::RequireCheckedSyscalls)
Your default perl library install directory is $ExtUtils::MakeMaker::Config{installsitelib}.
This works fine but has the drawback that the library files will not be found
anymore if the perl version gets upgraded. Since $site_lib is in
your default library search path I can use that instead

EOT
            if !$opt_y;
        # arch directories needed to get packlist into /usr/local/lib/site_perl
        map $ExtUtils::MakeMaker::Config{$_} = $site_lib, qw(installsitearch installsitelib sitelibexp sitearchexp) if
            $opt_y ||
            prompt("Use $site_lib as default library install directory ?", "y") =~ /^y/i;
    }

    print <<EOT	## no critic (UselessNoCritic InputOutput::RequireCheckedSyscalls)
This package comes with a program that I can try
to install in $ExtUtils::MakeMaker::Config{installsitebin}.

EOT
if !$opt_y;
    push @programs_to_install, "bin/dhcp_test" if
        $opt_y ||
        prompt("Install dhcp_test, a utility to test DHCP servers ?", "y") =~ /^y/i;
}

my $option_file = "$FindBin::Bin/t/options.$^O";
my $new = "$option_file.new.$$";
open(my $fh, ">", $new) || die "Could not open '$new': $!";
printf($fh "KEEP=0\nSTRACE=0\n") ||
    die "Error writing to '$new': $!";
eval {
    close($fh) || die "Could not close '$new': $!";
    rename($new, $option_file) ||
        die "Could not rename '$new' to '$option_file': $!";
};
if ($@) {
    $fh = undef;	# close file if open
    unlink($new) || die "Could not unlink '$new': $! after $@";
    die $@;
}

settings();

WriteMakefile
    (NAME		=> 'DHCP::Test',
     # finds $VERSION, requires EU::MM from perl >= 5.5
     VERSION_FROM	=> 'lib/DHCP/Test/Package.pm',
     eval {
         ## no critic (UselessNoCritic MagicNumbers)
         ExtUtils::MakeMaker->VERSION(6.55_01);
         1;
     } ? (BUILD_REQUIRES	=> {
         # To generate README.md when editing source. Not needed for build
         # "Pod::Markdown::Github"	=> "0.02",
         # Only for the tests
         "Test::More"	=> "0.01",
     }) : (),
     PREREQ_PM		=> {
         "Exporter::Tidy"		=> "0.06",
         # After OpenBSD segfault patches (2014 should be old enough)
         # We actually only directly use subpackage IO::Interface::Simple
         # but internally that uses IO::Interface anyways
         "IO::Interface"	=> "1.08",
         "Time::HiRes"		=> "0.01",
     },
     ABSTRACT		=> 'Ton Hospel <DHCP-Test@ton.iguana.be>',
     AUTHOR		=> 'Ton Hospel <ton@>',
     PL_FILES		=> {},
     EXE_FILES		=> \@programs_to_install,
     $^O eq "MSWin32" ? (
         PM_FILTER	=> '$(PERL) -p -e1',
     ) : (),
     clean		=> {
         FILES => '$(DISTNAME).ppd ppm cover_db nytprof nytprof.out* t/options t/options.* md5-versions.old *.bak',
     },
     #Value must be from legacy list of licenses here
     #http://search.cpan.org/perldoc?Module%3A%3ABuild%3A%3AAPI
     LICENSE		=> 'perl',
 );

# START MY
# autogenerated by release_pm
BEGIN {
    package MY;

    use vars qw(%postamble);

    $postamble{ppm} = <<'EOT';
ppm: $(DISTNAME).ppd
	makeppd.pl "--perl=$(PERLRUN)" --min_version=1.016 "--zip=$(ZIP)" "--tar=$(TAR)" "--compress=$(COMPRESS)" --leave=ppm "--objects=$(OBJECT)" $(DISTNAME).ppd $(VERSION)
EOT

    $postamble{ppm3} = <<'EOT';
ppm3: $(DISTNAME).ppd
	makeppd.pl "--perl=$(PERLRUN)" --min_version=1.016 "--zip=$(ZIP)" "--tar=$(TAR)" "--compress=$(COMPRESS)" --ppm_version=3 --leave=ppm3 "--objects=$(OBJECT)" $(DISTNAME).ppd $(VERSION)
EOT

    $postamble{ppm4} = <<'EOT';
ppm4: $(DISTNAME).ppd
	makeppd.pl "--perl=$(PERLRUN)" --min_version=1.016 "--zip=$(ZIP)" "--tar=$(TAR)" "--compress=$(COMPRESS)" --ppm_version=4 --leave=ppm4 "--objects=$(OBJECT)" $(DISTNAME).ppd $(VERSION)
EOT

    $postamble{ppd} = <<'EOT';
$(DISTNAME).ppd: all ppd
EOT

    $postamble{cover} = <<'EOT';
cover:
	cover -delete
	mkdir cover_db
	-HARNESS_PERL_SWITCHES=-MDevel::Cover make test
	cover
EOT

    $postamble{critic} = <<'EOT';
critic:
	perlcritic .
EOT

    $postamble{ppm_install} = <<'EOT';
ppm_install: ppm
	ppm install ppm/$(DISTNAME).ppd
EOT

    $postamble{ppm_uninstall} = <<'EOT';
ppm_uninstall:
	ppm uninstall $(DISTNAME)
EOT

    # You will need to add your own entry for README.md
    $postamble{commit} = <<'EOT';
commit: all README.md
	release_pm
	git commit -av
EOT

    $postamble{version_check} = <<'EOT';
all ::
	$(NOECHO) $(FULLPERL) version_check "$(NAME)" "$(DESTINSTALLSITELIB)" "$(VERSION)" "$(VERSION_FROM)"
EOT

undef &postamble;	## no critic (UselessNoCritic AmpersandSigils)
}
{
    package MY;
    sub postamble {
	return shift->SUPER::postamble() . join("\n", @postamble{sort {uc $a cmp uc $b || $a cmp $b } keys %postamble});
    }
}
# END MY

sub settings {
    package MY;
    our %postamble;

    $postamble{README} = <<'EOT';
README.md: bin/dhcp_test
	pod2readme -f github $< $@

dist: README.md
EOT
}
